### PR TITLE
Start Cypress browser with prefers-reduced-motion set to true

### PR DESCRIPTION
**Pull request by gusaiani**
One step to serve issue #17726 

Programatically sets this preference so that Cypress browser will honor it in e2e test environments.

In later PRs, as we remove or simplify animations whenever `prefers-reduced-motion` is `true`, e2e tests will reflect the changes.

## How to Test

Run Cypress locally in `open` mode.

Choose one or few tests to run in the browser.

1. Developer tools should open automatically just as they have been doing (this is another config that was already present in the same callback function, see diff in this PR)

2. Run the following line in the Cypress browser dev tools console:

```js
window.matchMedia('(prefers-reduced-motion: reduce)').matches;
```

console should output `true`.

### CODE REVIEW

**Comment by ariya** [comment]
If the setting is only applied to Firefox and Chrome/Chromium, will this create a discrepancy since the tests are executed in the CI using Electron instead? Imagine a developer working on a test running with Firefox on their laptop and the test passes just fine, but it still has the potential to fail on CI due to the different setting for Electron.

**Comment by gusaiani** [comment]
@ariya 

> If the setting is only applied to Firefox and Chrome/Chromium, will this create a discrepancy since the tests are executed in the CI using Electron instead?

tldr: It might. If this will mater or not is a case-by-case scenario.

My current understanding of this:

A media query can generate conditional CSS from the `prefers-reduced-motion` setting. 

~~~css
@media (prefers-reduced-motion) {
  someSelector {
    // any CSS, probably something to reduce animation
  }
}
~~~

Next, work needs to happen to treat animations, case-by-case, in CSS.

Each case may or may not affect e2e testing. 
The desired effect is speed gain.
I would venture a guess that, even for tests that cross animations, a majority of cases will have no undesired effect.

Getting to Electron, from the subset of tests that may be negatively affected at all by the discrepancy (besides being a bit slower), it seems it would be possible to treat that animation or e2e test and solve issues.

That said, do we rely on Electron e2e tests today? Is our app run inside Electron in any situation?

**Comment by nemanjaglumac** [comment]
> That said, do we rely on Electron e2e tests today? Is our app run inside Electron in any situation?

Whenever you run `yarn test-cypress-no-build` locally and in the CI, it runs inside Electron by default.

**Comment by gusaiani** [comment]
> Whenever you run yarn test-cypress-no-build locally and in the CI, it runs inside Electron by default.

@nemanjaglumac Thanks. I'm ignorant on this. What are the reasons? Just curious.

**Comment by nemanjaglumac** [comment]
> Thanks. I'm ignorant on this. What are the reasons? Just curious.

That's a good question for the Cypress product team :). That's the default browser they've chosen (probably for a good reason). It's possible to run tests in different headless browsers by providing `--browser` or `--b` flag, though.

**Comment by gusaiani** [comment]
Thanks for the answer, @nemanjaglumac 

So if the following assumptions are correct:

* our app is only run in Electron for e2e tests
* e2e tests now running in Electron can be run in Chrome headless or other
* we are set to gain by doing work around `prefers-reduced-motion`

any objections to updating the e2e run scripts so that they use Chrome instead of Electron?

cc @ariya @flamber

**Comment by nemanjaglumac** [comment]
Just found this doc:
https://docs.cypress.io/guides/guides/launching-browsers#Chrome-Browsers

> The Electron browser has the advantage of coming baked into Cypress and does not need to be installed separately.

> By default, when running cypress run from the CLI, we will launch all browsers headlessly.

For Chrome:
`cypress run --browser chrome`

> To use this command in CI, you need to install the browser you want - or use one of our docker images.

> By default, we will launch Chrome in headlessly during cypress run.

**Comment by gusaiani** [comment]
@nemanjaglumac gotcha

So, not to completely negate any time saving from this endeavor:

do you happen to know if Chrome headless can easily be cached in our CI pipeline?

**Comment by nemanjaglumac** [comment]
> do you happen to know if Chrome headless can easily be cached in our CI pipeline?

Tbh, I don't know :/
@paoliniluis can you please shed some light on this?

**Comment by paoliniluis** [comment]
yup, we do (in fact, we don't, it's inherited by the circleci image that we use as a base) install chrome in the images we use for our CI: layer 52 of this container image: https://hub.docker.com/layers/circleci/clojure/openjdk-11-buster-node-browsers/images/sha256-67bf09b60ce2381e263796d2e3bb05902b82350877662c14824f6c695c9a5584?context=explore

you have firefox and chrome both installed already, so if you switch to the new mode, it should work straight out of the box

**Comment by gusaiani** [comment]
Thank you @paoliniluis! 
Do you happen to "just know"TM where to change the lines to run with the flag in CI?

Just to save head bouncing time.

**Comment by paoliniluis** [comment]
yeah, try passing the argument to line 958 in the config.yml in .circleci folder, that should do the trick. Also another trick: if something fails, remember that you can get into the container in the step that failed with SSH (CircleCI does a very neat port forwarding to the containers)

**Comment by gusaiani** [comment]
Thank you @paoliniluis, trying this 🙇

**Comment by ariya** [comment]
> any objections to updating the e2e run scripts so that they use Chrome instead of Electron?

No objection but in that case, let's have a separate PR that switches from Electron to Chrome first and let's watch it over a few days or so to ensure that it won't introduce any flakiness. One advantage of the use of Electron to Cypress is its guaranteed speed (faster than controlling external browsers) and hence we want to exclude any timing-related uncertainties.

Also, something we need to keep in mind when switching to execute Cypress tests with an external browser: now the CI run and local development **can** be different. Minimizing e.g. the version difference is possible if the browser inside Circle CI container image is updated (1) periodically every new browser release (6 weeks or so), or (2) whenever we detect the divergence. In either case, the cost can be small but still non zero.

**Comment by gusaiani** [comment]
@ariya 

Fair points, thank you.

Given everything we've seen today, notably the difficulty getting the tests to even use anything other than Electron, plus all the considerations above.

Comparing that with the eventual gains in test speed we may have.

It's starting to look like we should not go forward with this for now, and focus on higher return initiatives.

As @flamber gave the suggestion to try this, I wonder especially about what he has to say. But also of course whoever may have an opinion.

**Comment by flamber** [comment]
Seems like there's a hacky workaround: https://dev.to/ecosia/clicking-stuff-in-e2e-tests-smooth-scrolling-electron-flags-and-cypress-2a1c#electron-doesnt-support-that
While reducing motion can certainly speed up tests, my main priority was accessibility. Of course we want to first be able to test this, so it makes sense to initially be able to do this in Cypress.

**Comment by gusaiani** [comment]
👍, let's see how it behaves with the Electron flag in the script call.

**Comment by gusaiani** [comment]
@nemanjaglumac (and whoever may be interested), we would benefit from your take on this.

I added an env var with the script calls (the ones that are "no-build"), to make Electron receive the "prefers-reduced-motion" flag.

Nothing new broke, tests continued to pass.
But at that point, I had no evidence that the flags reached Electron.

To check that Electron gets the flag, I added a conditional in the implementation code, that would throw an error (tests would fail) if the flag was set to true.

This seems to happen as expected in this [commit](https://github.com/metabase/metabase/pull/18041/commits/a66f509f0729e68571b24713feefb7ec7def24ce), as all e2e tests failed.

Then I removed this conditional and, again, tests should pass.

Do you think this is robust enough as a demonstration that the flag reaches Electron? 
Any other ideas/comments?

**Comment by nemanjaglumac** [comment]
@gusaiani that looks good (and like a smart check) to me. I'm not sure what else could we check at this moment. We can always disable this (revert) if we figure out flakes are increasing.

Just a side note/question - Have you tried comparing the total running time before and after (either for the individual spec or maybe for a specific folder)?

**Review by nemanjaglumac — APPROVED** [review]
👏
Let's try this 🚀

**Review by flamber — APPROVED** [review]
Since this is just step1, which just enabled the possibility, then let's see what speed can come from the actual reduced animations in other PRs.

**Comment by gusaiani** [comment]
Thank you fellows 🤜❤️🤛

**Comment by gusaiani** [comment]
@nemanjaglumac 

> Just a side note/question - Have you tried comparing the total running time before and after (either for the individual spec or maybe for a specific folder)?

Yes, we still need to simplify animations when flag is "true" to get any possible speed gains 👍
